File: samples/AdventureWorks/AdventureWorks.Client.Common/DataObjects/Person/AddressObjectCustomized.cs
using System;
using Xomega.Framework;
using Xomega.Framework.Properties;

namespace AdventureWorks.Client.Objects
{
    public class AddressObjectCustomized : AddressObject
    {
        public AddressObjectCustomized()
        {
        }

        public AddressObjectCustomized(IServiceProvider serviceProvider) : base(serviceProvider)
        {
        }

        // construct properties and child objects
        protected override void Initialize()
        {
            base.Initialize();
            // add custom construction code here
        }

        // perform post intialization
        protected override void OnInitialized()
        {
            base.OnInitialized();
            AddressIdProperty.Change += OnAddressChanged;
        }

        private void OnAddressChanged(object sender, PropertyChangeEventArgs e)
        {
            if (e.Change.IncludesValue() && !DataProperty.Equals(e.OldValue, e.NewValue))
            {
                Header addr = AddressIdProperty.Value;
                AddressLine1Property.SetValue(addr == null ? null : addr[Enumerations.BusinessEntityAddress.Attributes.AddressLine1]);
                AddressLine2Property.SetValue(addr == null ? null : addr[Enumerations.BusinessEntityAddress.Attributes.AddressLine2]);
                CityStateProperty.SetValue(addr == null ? null : addr[Enumerations.BusinessEntityAddress.Attributes.City]
                                                        + ", " + addr[Enumerations.BusinessEntityAddress.Attributes.State]);
                PostalCodeProperty.SetValue(addr == null ? null : addr[Enumerations.BusinessEntityAddress.Attributes.PostalCode]);
                CountryProperty.SetValue(addr == null ? null : addr[Enumerations.BusinessEntityAddress.Attributes.Country]);
            }
        }

    }
}

File: samples/AdventureWorks/AdventureWorks.Client.Spa/DataObjects/Person/AddressObjectCustomized.ts
import { BusinessEntityAddress } from 'Enumerations/Enumerations';
import { Header } from 'xomega';

import { AddressObject as GeneratedDataObject } from 'DataObjects/Person/AddressObject';

export class AddressObject extends GeneratedDataObject {

    // perform post intialization
    onInitialized() {
        super.onInitialized();
        let obj = this;
        this.AddressId.InternalValue.subscribe(function (h: Header) {
            obj.AddressLine1.InternalValue(h ? h.attr[BusinessEntityAddress.Attributes.AddressLine1] : null);
            obj.AddressLine2.InternalValue(h ? h.attr[BusinessEntityAddress.Attributes.AddressLine2] : null);
            // for composite properties display null if an underlying property is null
            obj.CityState.InternalValue(h && h.attr[BusinessEntityAddress.Attributes.City] ?
                `${h.attr[BusinessEntityAddress.Attributes.City]}, ${h.attr[BusinessEntityAddress.Attributes.State]}` : null);
            obj.PostalCode.InternalValue(h ? h.attr[BusinessEntityAddress.Attributes.PostalCode] : null);
            obj.Country.InternalValue(h ? h.attr[BusinessEntityAddress.Attributes.Country] : null);
        });
    }
}

